### ISSUE TITLE
Diverging subexpression lint should not fire on `todo!()`

### ISSUE THREAD

**Issue by digama0** [rust-lang/rust-clippy#10243]
### Summary

For incomplete code, it is not that helpful to point out that a subexpression diverges. I suggest suppressing the lint warning when the diverging subexpression is an application of the `todo!()` macro.

### Lint Name

diverging_sub_expression

### Reproducer

I tried this code:

```rust
#![allow(unused)]

fn main() {
  let x: u32 = todo!();
  println!("{x}");
}
```

I saw this happen:

```
warning: sub-expression diverges
 --> src/main.rs:4:16
  |
4 |   let x: u32 = todo!();
  |                ^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#diverging_sub_expression
  = note: `#[warn(clippy::diverging_sub_expression)]` on by default
  = note: this warning originates in the macro `todo` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I expected to see this happen: No warning

### Version

```text
rustc 1.68.0-nightly (3020239de 2023-01-09)
binary: rustc
commit-hash: 3020239de947ec52677e9b4e853a6a9fc073d1f9
commit-date: 2023-01-09
host: x86_64-unknown-linux-gnu
release: 1.68.0-nightly
LLVM version: 15.0.6
```

**Comment by gahag-cw** [comment]
Also `unimplemented!`.

**Comment by Centri3** [comment]
`unimplemented!` doesn't make the same guarantee as `todo!`, in that it's not "unfinished", it's just unimplemented. I think it makes sense to continue linting it, however rare it may be.

**Comment by dpc** [comment]
I'm trying to update Rust to 1.72.0 in our project, and now tons of code where we have `unimplemented!()` triggers this warning. Seems to me due to `#[async_trait]`?

**Comment by Centri3** [comment]
While your issue has been solved, that could be a sign we need to check if the expression is from a procedural macro (using `is_from_proc_macro`).

**Comment by mtelahun** [comment]
Did this recently make it into rust-stable? I just recently updated my -stable tool-chain and I am being affected by this issue wherever I use the todo! macro. This is extremely inconvenient for me because I have my pre-commit hooks run clippy in all my rust projects, which prevents me from committing any changes that include the macro. For the time-being I have disabled this lint at the crate level but that's not a viable long-term solution.

**Comment by digama0** [comment]
I don't think this is a new issue; I reported it over a year ago and as far as I know it affects all versions of clippy containing the lint. I don't believe there are any fixes in the pipeline.
